File: ruby/lib/basket_actions/add_to_basket_command.rb
require_relative '../user_interface/usage_exception'

module BasketActions

	class AddToBasketCommand

		def initialize(basket, catalogue, warehouse)
			@basket = basket
			@catalogue = catalogue
			@warehouse = warehouse
		end

		def run(cmd)
			if cmd.args.length > 3 || cmd.args.length < 2
				$stderr.puts "ERROR: Usage: a sku [numitems]"
				return
			end
			sku = cmd.arg(1)
			num_items = 1
			if cmd.args.length == 3
				begin
					num_items = Integer(cmd.arg(2))
				rescue ArgumentError
					raise UserInterface::UsageException.new("The number of items must be a number greater than zero.")
				end
			end
			if num_items <= 0
				raise UserInterface::UsageException.new("The number of items must be a number greater than zero.")
			end
			sku_code = @catalogue.lookup_code(sku)
			raise UserInterface::UsageException.new("Product " + sku + " unknown.") if sku_code.nil?
			@warehouse.must_stock(sku_code, num_items)
			@basket.add(sku, @catalogue, num_items)
		end

	end
end

File: python3/basket_actions/add_to_basket_command.py
from user_interface.usage_exception import UsageException

class AddToBasketCommand:

    def __init__(self, basket, catalogue, warehouse):
        self.basket = basket
        self.catalogue = catalogue
        self.warehouse = warehouse

    def run(self, cmd):
        if len(cmd.args) > 3 or len(cmd.args) < 2:
            print("ERROR: Usage: a sku [numitems]")
            return
        sku = cmd.arg(1)
        numItems = 1
        if len(cmd.args) == 3:
            try:
                numItems = int(cmd.arg(2))
            except:
                raise UsageException("The number of items must be a number greater than zero.")
        if numItems <= 0:
            raise UsageException("The number of items must be a number greater than zero.")
        skuCode = self.catalogue.lookupCode(sku)
        if not skuCode:
            raise UsageException("Product " + sku + " unknown.")
        self.warehouse.mustStock(skuCode, numItems)
        self.basket.add(sku, self.catalogue, numItems)

